### PR TITLE
java: support for linking jar files

### PR DESCRIPTION
**Pull request by noverby**

### CODE REVIEW

**Comment by jpakkane** [comment]
Based on CI and my tests, this code as it stands does not actually pass. Did this work for you? If not, please write a description of known issues in the pull request description. Otherwise people have to spend a lot of time debugging why things don't work.

In Meson the concept of `is_linkable` relates to "classical" C style linking that is done with the system linker. Changing jar targets to be linkable should not be done, because then you can try to link a jar with a regular library or vice versa. Please remove that change. Instead in the Jar object you should verify all objects that were passed in the `links_with` keyword and abort with a descriptive error if any of them is not a Jar object.

File: mesonbuild/build.py [hunk 1]
```diff
@@ -1902,6 +1902,9 @@ def __init__(self, name, subdir, subproject, is_cross, sources, objects, environ
         for s in self.sources:
             if not s.endswith('.java'):
                 raise InvalidArguments('Jar source %s is not a java file.' % s)
+        for t in self.link_targets:
+            if not isinstance(Jar, t):
+                raise InvalidArguments('Link target %s is not a jar target.' % t)
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[Flake8]__

[W291] trailing whitespace

**Comment by noverby** [comment]
But `is_linkable` if true for C# libraries, shouldn't it be true for Jar libraries too? If it is set to false it prevents Meson from linking both C# and Jar Libraries.

**Comment by noverby** [comment]
See [link method](https://github.com/mesonbuild/meson/blob/master/mesonbuild/build.py#L927) in build.py.

**Comment by jpakkane** [comment]
Hmmm. Okay, I guess we can go with that, we can fix it later. But why does the test not run? I tried making the class `public` but it did not help (they are in the same namespace so it should not matter, but just to be sure).

File: mesonbuild/build.py [hunk 2]
```diff
@@ -1918,6 +1921,9 @@ def get_java_args(self):
     def validate_cross_install(self, environment):
         # All jar targets are installable.
         pass
+
```
Comment (ignatenkobrain) [inline mesonbuild/build.py]: __[Flake8]__

[W293] blank line contains whitespace

**Comment by jpakkane** [comment]
Do you have any idea why the jar linking does not seem to work?  Everything seems to be ok.

**Comment by noverby** [comment]
When Meson runs the jar files, it needs to use the classpath arguments. This is not the case right now in Meson.

File: mesonbuild/compilers/java.py [hunk 3]
```diff
@@ -80,7 +80,7 @@ def get_pch_name(self, header_name):
 
     def get_buildtype_args(self, buildtype):
         return java_buildtype_args[buildtype]
-
+
```
Comment (ignatenkobrain) [inline mesonbuild/compilers/java.py]: __[Flake8]__

[W293] blank line contains whitespace

**Comment by jpakkane** [comment]
I looked through Java documentation while debugging this earlier and I noticed that Jar files actually support something like `rpath` in their manifest files.

The real, proper solution to this would then to exactly the same thing for jars as what we do for executables currently. That is, write a "pseudo rpath" entry to the manifest file so the jars can be run directly with `java -jar foobar.jar` and then remove this setting during install (and allow setting `install_rpath` in them for extra bonus points).

This is a fair bit of work, though, so if you don't want to do that now we can get this in first to get other stuff unblocked and fix that sometime later.

**Comment by noverby** [comment]
We should go for the proper solution right away.

**Comment by noverby** [comment]
New pull request using manifest:
https://github.com/mesonbuild/meson/pull/3763

**Pull request closed without merging**